src/main/kotlin/solver/unconstrained/AltaLineSearch.kt

```diff
@@ -1,19 +1,48 @@
 package autodiff.solver.unconstrained
-//
-//import autodiff.*
-//import autodiff.operator.*
-//
-//class AltaLineSearch(private val function: Expression) {
-//    fun solveApproximateMinimum(initialGuess: VariableMap, direction: VariableMap): Double {
-//        val m: Double = function.solveGradient(initialGuess).dot(direction) // directional derivative
-//        var alpha = Variable("alpha")
-//        val beta = 0.1 // constant on [0,1], always positive
-//        val tau = 0.9 // constant on [0,1], always positive
-//        val f_0 = function.evaluate(initialGuess) // f_k
-//        for (k in 1..100) {
-//            var coef = function.forwardAutoDiff(alpha, )
-//            alpha -= (coef.get(1)) / (2 * coef.get(2))
-//        }
-//        return alpha
-//    }
-//}
\ No newline at end of file
+
+import autodiff.*
+import autodiff.operator.*
+
+class AltaLineSearch(private val function: Expression) {
+    /**
+     * Optimizes the step size "alpha" for gradient descent to minimize the function f(x, y, z, ...)
+     * by iterating with Newton's method using the derivatives df/dalpha & d^2f/dalpha^2
+      */
+    fun solveApproximateMinimum(initialGuess: VariableMap): VariableMap {
+        var iteration = initialGuess
+        var alpha = Variable("alpha")
+        // Arbitrary number of iterations - the terminating condition should depend on some sort of convergence estimate
+        for (k in 1..200) {
+            val gradient = function.solveGradient(iteration)
+            var alphaMap = HashMap<Variable, Expression>()
+            for (variable in gradient.map.keys) {
```

Comment:
If you'd like, `for((variable,value) in gradient.map) {`, but honestly whatever

Reply:
oh actually that's not true, you could do `for((variable, gradientValue) in gradient.map) {` and then use the `gradientValue` variable instead of `gradient.get(variable)`

Reply:
but either way the code is perfectly readable so /shrug whatever